### PR TITLE
Implement clean uninstall of operator

### PR DESCRIPTION
**Pull request by VaishnaviHire**
This commit includes changes in the controller to
remove resources created in the namespaces other than
the operator namespace. Removal of external resources
is triggered once a delete configMap is added to the
operator namespace

### CODE REVIEW

**Comment by nakfour** [comment]
@VaishnaviHire how do we test this?

**Comment by VaishnaviHire** [comment]
https://github.com/opendatahub-io/opendatahub-operator/pull/117#issuecomment-819564177

I was able to test this using following steps:
1. Deploy custom operator image
2.  Create KfDef using odh-deployer script, so that we have `monitoring` and `applications` namespaces are set up.
3. Once the operator is reconciled successfully, create a ConfigMap in `operatorNamespace` with label  `api.openshift.com/addon-managed-odh-delete` : `true` . This should trigger reconcile that delete the resources.
4.  To test deletion of csv, I created a dummy csv object, before creating the delete configMap.

File: pkg/controller/kfdef/kfdef_controller.go [hunk 1]
```diff
@@ -437,3 +494,100 @@ func setAnnotations(configPath string, annotations map[string]string) error {
...
+	// Get watchNamespace
+	operatorNamespace, err := k8sutil.GetOperatorNamespace()
+	if err != nil {
+		return err
+	}
+
+	// Delete generated namespaces
+	generatedNamespaces := &v1.NamespaceList{}
+	nsOptions := []client.ListOption{
+		client.MatchingLabels{rhodsGeneratedNamespaceLabel: "true"},
```
Comment (crobby) [inline pkg/controller/kfdef/kfdef_controller.go:535]: I'm not seeing this label anywhere on the projects/namespaces....for the redhat-ods-monitoring namespace, I see the following labels...
    hive.openshift.io/managed: "true"
    olm.operatorgroup.uid/b4e6a5a4-779a-4145-a6fc-403343705480: ""
    openshift.io/cluster-monitoring: "true"

Comment (crobby) [reply pkg/controller/kfdef/kfdef_controller.go:535]: Oh wait...I see that the deployer is expected to add those labels...I do not have that running with that change yet.

Comment (VaishnaviHire) [reply pkg/controller/kfdef/kfdef_controller.go:535]: The change to add labels recently merged in odd-deployer : https://github.com/red-hat-data-services/odh-deployer/pull/84 

Did you pull in latest commit?

Comment (crobby) [reply pkg/controller/kfdef/kfdef_controller.go:535]: Yes...I see now.  Totally just me confusing myself.

Comment (vpavlin) [reply pkg/controller/kfdef/kfdef_controller.go:535]: It should not be prefixed `rhods`

File: pkg/controller/kfdef/kfdef_controller.go [hunk 2]
```diff
@@ -35,6 +42,11 @@ const (
 	finalizer = "kfdef-finalizer.kfdef.apps.kubeflow.org"
 	// finalizerMaxRetries defines the maximum number of attempts to add finalizers.
 	finalizerMaxRetries = 10
+	// deleteConfigMapLabel is the label for configMap used to trigger operator uninstall
+	// TODO: Label should be updated if addon name changes
+	deleteConfigMapLabel = "api.openshift.com/addon-managed-odh-delete"
+	// rhodsGeneratedNamespaceLabel is the label added to all the namespaces genereated by odh-deployer
+	rhodsGeneratedNamespaceLabel = "opendatahub.io/generated-namespace"
```
Comment (crobby) [inline pkg/controller/kfdef/kfdef_controller.go:49]: I think we would also want the variable name to be more like "odhGeneratedNamespaceLabel"

**Pull request merged**